compiler/src/compiler/mir.rs

```diff
@@ -0,0 +1,588 @@
+        Self(id)
+    }
+    fn to_usize(&self) -> usize {
+        self.0
+    }
+}
+
+impl Body {
+    pub fn new() -> Self {
+        Self {
+            expressions: vec![],
+        }
+    }
+
+    pub fn push(&mut self, id: Id, expression: Expression) {
+        self.expressions.push((id, expression));
+    }
+    pub fn push_with_new_id(
+        &mut self,
+        id_generator: &mut IdGenerator<Id>,
+        expression: Expression,
+    ) -> Id {
+        let id = id_generator.generate();
+        self.push(id, expression);
+        id
+    }
+    pub fn insert_at_front(&mut self, expressions: Vec<(Id, Expression)>) {
+        let old_expressions = mem::take(&mut self.expressions);
+        self.expressions.extend(expressions);
+        self.expressions.extend(old_expressions);
+    }
+    pub fn remove_all<F>(&mut self, mut predicate: F)
+    where
+        F: FnMut(Id, &Expression) -> bool,
+    {
+        self.expressions
+            .retain(|(id, expression)| !predicate(*id, expression));
+    }
+    pub fn sort_by<F>(&mut self, predicate: F)
+    where
+        F: FnMut(&(Id, Expression), &(Id, Expression)) -> Ordering,
+    {
+        self.expressions.sort_by(predicate);
+    }
+
+    pub fn return_value(&mut self) -> Id {
+        let (id, _) = self.expressions.iter_mut().last().unwrap();
+        *id
+    }
+
+    /// Flattens all `Expression::Multiple`.
+    pub fn flatten_multiples(&mut self) {
+        let old_expressions = mem::take(&mut self.expressions);
+
+        for (id, mut expression) in old_expressions.into_iter() {
+            if let Expression::Multiple(mut inner_body) = expression {
+                inner_body.flatten_multiples();
+                let returned_by_inner = inner_body.return_value();
+                for (id, expression) in inner_body.expressions {
+                    self.expressions.push((id, expression));
+                }
+                self.expressions
+                    .push((id, Expression::Reference(returned_by_inner)));
+            } else {
+                if let Expression::Lambda { body, .. } = &mut expression {
+                    body.flatten_multiples();
+                }
+                self.expressions.push((id, expression));
+            }
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (Id, &Expression)> {
+        self.expressions
+            .iter()
+            .map(|(id, expression)| (*id, expression))
+    }
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (Id, &mut Expression)> {
+        self.expressions
+            .iter_mut()
+            .map(|(id, expression)| (*id, expression))
+    }
+    pub fn into_iter(self) -> impl Iterator<Item = (Id, Expression)> {
+        self.expressions.into_iter()
+    }
+}
+
+impl Body {
+    pub fn visit(&mut self, visitor: &mut dyn FnMut(Id, &mut Expression, bool)) {
+        self.visit_rec(visitor);
+    }
+    fn visit_rec(&mut self, visitor: &mut dyn FnMut(Id, &mut Expression, bool)) {
```

Comment:
Can we merge these methods since they have the same signature?

```suggestion
    pub fn visit(&mut self, visitor: &mut dyn FnMut(Id, &mut Expression, bool)) {
```